server/swap/swap.go

```diff
@@ -823,11 +823,31 @@ func (s *Swapper) processAck(msg *msgjson.Message, acker *messageAcker) {
 	} else {
 		tAckMS := encode.UnixMilli(tAck)
 		if acker.isMaker {
-			s.authMgr.RecordCompletedOrder(acker.user, acker.match.Maker.ID(), tAck)
+			// A maker order is complete if it has force immediate or no
+			// remaining amount, indicating it is off the books.
+			lo := acker.match.Maker
+			if lo.Remaining() == 0 || lo.Force == order.ImmediateTiF {
+				s.authMgr.RecordCompletedOrder(acker.user, acker.match.Maker.ID(), tAck)
+			}
+
 			acker.match.Sigs.MakerRedeem = sigBytes
 			s.storage.SaveRedeemAckSigA(mktMatch, sigBytes, tAckMS)
 		} else {
-			s.authMgr.RecordCompletedOrder(acker.user, acker.match.Taker.ID(), tAck)
+			// A taker order is complete if it is one of: (1) a market order,
+			// which does not go on the books, (2) limit with force immediate,
+			// or (3) limit with no remaining amount.
+			var complete bool
+			switch ot := acker.match.Taker.(type) {
+			case *order.LimitOrder:
+				complete = ot.Force == order.ImmediateTiF || ot.Remaining() == 0
+			case *order.MarketOrder:
+				complete = true
+			}
+
+			if complete {
+				s.authMgr.RecordCompletedOrder(acker.user, acker.match.Taker.ID(), tAck)
+			}
+
```

Comment:
Only the market, who runs the book and performed matching, knows if the order is actually on the book, but these criteria should exclude orders that are (still) on the book from being recorded as completed.